pkg/instance/instance.go

```diff
@@ -536,3 +552,39 @@ func RunnerCmd(prog, fwdAddr, os, arch string, poolIdx, vmIdx int, threaded, new
 	return fmt.Sprintf("%s -addr=%s -os=%s -arch=%s -pool=%d -vm=%d "+
 		"-threaded=%t -new-env=%t", prog, fwdAddr, os, arch, poolIdx, vmIdx, threaded, newEnv)
 }
+
+type Semaphore struct {
+	count int
+	ch    chan struct{}
+}
+
+func NewSemaphore(count int) *Semaphore {
+	s := &Semaphore{
+		count: count,
+		ch:    make(chan struct{}, count),
+	}
+	for i := 0; i < count; i++ {
+		s.Signal()
+	}
+	return s
+}
+
+func (s *Semaphore) Wait() {
+	<-s.ch
+}
+
+func (s *Semaphore) WaitC() <-chan struct{} {
+	return s.ch
+}
+
+func (s *Semaphore) Available() int {
+	return len(s.ch)
+}
+
+func (s *Semaphore) Signal() {
+	s.ch <- struct{}{}
+	if av := s.Available(); av > s.count {
```

Comment:
This will never fire, len can't be larger than cap.
You can check that `len(ch) != cap(ch) ` _before_ doing the send.